Adding Kill(bool entireProcessTree) -- terminates a process tree

bgribaudo opened the pull request:
[Implements API approved by #26234]

## Summary
When `Kill(bool entireProcessTree)`'s argument is false, behaves identical to `Kill()`; when true, terminates the current process and all descendant processes (children, grandchildren, etc.).

## Platform Implementation Status
Implemented for Windows, OSX and Linux; not implemented for UAP, UnknownUnix or FreeBSD. 

## Notes
Potential low-hanging fruit: Behind-the-scenes, the logic in this PR enumerates child processes. This functionality could be leveraged to easily implement #25855. 

Windows-specific logic inspired by [msbuild's NativeMethodsShared KillTree](https://github.com/Microsoft/msbuild/blob/0591c15d6c638cad38091fbe625dde968f86748d/src/Shared/NativeMethodsShared.cs#L1068).

[This PR replaces #31827]

danmosemsft commented:
@dotnet-bot test Linux arm release build (IO error)

danmosemsft commented:
@MattGal who do we normally ask here when Tizen images are broken? The group name I was told last time did not work then.

danmosemsft commented:
@bgribaudo looks like merge conflicts 😸

danmosemsft commented on src/System.Diagnostics.Process/src/System/Diagnostics/Process.NonUap.cs:
```diff
@@ -0,0 +1,84 @@
+            return false;
+        }
+
+        private bool SafePredicateTest(Func<bool> predicate, bool fallbackResult = false)
```
Nit, since you never use fallbackResult, you could remove the parameter and assume false.

Don't need to reset PR for this though

bgribaudo replied:
Thanks, @danmosemsft! Parameter removed.

danmosemsft commented on src/System.Diagnostics.Process/src/System/Diagnostics/Process.OSX.cs:
```diff
@@ -94,6 +94,20 @@ public TimeSpan UserProcessorTime
+            {
+                EnsureState(State.HaveNonExitedId);
+
+                // Returns null if permissions do not allow querying the specified process
```
Seems like the Windows version of ParentProcessId throws Win32Exception when permissions are insufficient, whereas the OSX one returns null. I guess that's OK as they're private.

bgribaudo replied:
Thanks, @danmosemsft! For consistency, changed to throw an exception instead of return null.

bgribaudo commented:
@danmosemsft, I used the online resolve conflicts tool to resolve those conflicts. In the process, it merged master. Will that cause problems? (Do I need to replace that merge with a rebase or something like that?)

danmosemsft commented:
> Will that cause problems? 

@bgribaudo nope, but I only use that online tool when the merge is trivial, because waiting 20+ mins for CI is a time consuming way to discover that I didn't merge correctly.

danmosemsft commented:
Test failure on Nano

```
Xunit.Sdk.TrueException
Windows.10.Nano.Amd64.Open-x64-Debug
Get Repro environment
Unhandled Exception of Type Xunit.Sdk.TrueException
Message :
TestGetProcesses001 failed
Expected: True
Actual:   False
Stack Trace :
   at System.Diagnostics.Tests.ProcessTests.TestGetProcesses() in D:\j\workspace\windows-TGrou---74aa877a\src\System.Diagnostics.Process\tests\ProcessTests.cs:line 1014
```

danmosemsft commented:
No idea why it can't find itself. Might be worth adding logging of processes found, in the case this when this test fails.

bgribaudo commented:
@dotnet-bot test please

bgribaudo commented:
@danmosemsft, could you mark me as an assignee on this pull request? I think I need to be one in order to send @dotnet-bot commands. Thanks!

bgribaudo commented:
> No idea why it can't find itself. Might be worth adding logging of processes found, in the case this when this test fails.

@danmosemsft, I'm puzzled by this one...especially as it doesn't seem to have anything to do with code I've modified. I can try to take a look....

danmosemsft commented:
Dotnet bot shoukd accept orders from anybody...

@dotnet-bot test windows x64 debug build

MattGal commented:
@danmosemsft :  neither @mmitche nor I know about the Tizen stuff but @gbalykov might?

krwq commented:
@bgribaudo I think if you just say `test <copy paste of configuration name>` (include only that in the message) it should work. Try `@dotnet-bot help` if it will respond

bgribaudo commented:
Thanks, @krwq!

bgribaudo commented:
@dotnet-bot test windows x64 debug build

bgribaudo commented:
@dotnet-bot test windows x64 debug build

danmosemsft commented on src/Common/src/Interop/Windows/NtDll/Interop.NtQueryInformationProcess.cs:
```diff
@@ -0,0 +1,37 @@
+        };
+
+        [StructLayout(LayoutKind.Sequential)]
+        internal struct PROCESS_BASIC_INFORMATION
```
Nit, although I see pinvoke.net somehow shows all the members, some of these are not officially documented. The public headers define it as 
```
typedef struct _PROCESS_BASIC_INFORMATION {
    PVOID Reserved1;
    PPEB PebBaseAddress;
    PVOID Reserved2[2];
    ULONG_PTR UniqueProcessId;
    PVOID Reserved3;
} PROCESS_BASIC_INFORMATION;
```
We try to match public headers as closely as possible, so we would have
```
        public IntPtr Reserved1; // or UIntPtr
        public IntPtr PebBaseAddress;
        public IntPtr Reserved2a;
        public IntPtr Reserved2b;
        public IntPtr BasePriority;
        public UIntPtr UniqueProcessId;
        public UIntPtr Reserved3;
```
Some of this is a matter of taste I think, eg., `void*` would also work, but we don't seem to use that as much. I'm not sure of the general preference.

danmosemsft replied:
I'll fix it.

danmosemsft replied:
But we are relying on InheritedFromUniqueProcessId which is among the undocumented fields. Although, it is "documented" all over the web. It seems the documented way is to enumerate with the ToolHelp32 API's - ugh.

danmosemsft replied:
Never mind, it is documented in the DDK

bgribaudo commented:
@dotnet-bot test Linux arm64 release build

bgribaudo commented:
@dotnet-bot test osx x64 debug build
@dotnet-bot test Linux arm64 release build

bgribaudo commented:
@dotnet-bot test Linux arm64 release build

bgribaudo commented:
>>No idea why it can't find itself. Might be worth adding logging of processes found, in the case this when this test fails.
> @danmosemsft, I'm puzzled by this one...especially as it doesn't seem to have anything to do with code I've modified. I can try to take a look...

Looks like this failure is resolved. I added logging (to console) code and it went away. I removed the logging code and it stayed away. Hm...maybe something else was going on in master at the same time that caused the problem.

bgribaudo commented:
@dotnet-bot test Packaging All Configurations x64 Debug Build

danmosemsft commented:
ARM seems broken judging by previous
https://ci3.dot.net/job/dotnet_corefx/job/master/job/linux-TGroup_netcoreapp+CGroup_Release+AGroup_arm64+TestOuter_false_prtest/

Need to wait for folks to come back who know how to fix it.

danmosemsft commented:
@dotnet/dnceng could you please look at why all ARM64 test jobs are timing out (after 12 hours). are machines offline?

maririos commented:
It looks like people are OOF on vacation this week.
FYI @ilyas1974 in case he is around

ilyas1974 commented:
The machines are registering as online in our tooling. Are other jobs processing through ci3? Could the problem be on the server side?

danmosemsft commented:
Yes, eg the Linux ARM Release Build leg is through ci3.

On the Jenkins side it seemingly submitted to Helix without error
https://ci3.dot.net/job/dotnet_corefx/job/master/job/linux-TGroup_netcoreapp+CGroup_Release+AGroup_arm64+TestOuter_false_prtest/9380/consoleText

Somehow need to look into Helix to see what happened. Perhaps this helps
```
  Creating container named 'build-dd98e7893cac4288b02dbc916619b3de' in storage account dotnetbuilddrops.
  Creating container named 'build-dd98e7893cac4288b02dbc916619b3de' in storage account dotnetjobresults.
  Writing /mnt/j/workspace/dotnet_corefx/master/linux-TGroup_netcoreapp+CGroup_Release+AGroup_arm64+TestOuter_false_prtest/netci-archived-logs/TestList.json.
  Writing /mnt/j/workspace/dotnet_corefx/master/linux-TGroup_netcoreapp+CGroup_Release+AGroup_arm64+TestOuter_false_prtest/netci-archived-logs/BuildComplete.json.
  Started Helix job: CorrelationId = 800ffdec-3204-424f-9f31-2bc540c7b3ff
```

ilyas1974 commented:
That was the missing piece. We removed helix from the systems in the Ubuntu.1804.Arm64.Open queue to convert them to the new configuration that was agreed to (hosts for docker based agents). Let me work on getting Helix back on these some of these systems to get jobs running against these systems. Will update this thread once systems are back online. 

@riarenas , are jobs backed up in the Ubuntu.1804.Arm64 queue as well or just the Ubuntu.1804.Arm64.Open queue?

danmosemsft commented:
@riarenas 

> I don't see any helix machines heartbeating for the Ubuntu.1804.Arm64.Open queue, which is the one the job is waiting for, and is quite backed up.

Shouldn't Helix return immediately with an error in such a case? It doesn't sound like it's waiting for the queue to become free or something that would benefit from a 12 hour wait.

riarenas commented:
Ubuntu.1804.Arm64 is not backed up right now, only the open queue.

ilyas1974 commented:
Adding @JpratherMS to this thread as he has all the details on the machines in question. He will update this thread once the systems are back online.

riarenas commented:
@danmosemsft Helix received the work items fine, and from its point of view, it is simply waiting for a machine to be available to process the work, the problem here is that there won't be any machines polling the queue until they get reconfigured.

I think the new Helix sender that we use for Azdo jobs has a more aggressive timeout for instances like these, but I'm not aware of anything similar in Jenkins.

JpratherMS commented:
I have added 10 machines back to the 1804 open queue, please let me know if you see any further issues

danmosemsft commented:
thanks @JpratherMS 

@dotnet-bot test Linux arm64 Release Build please

bgribaudo commented:
@dotet-bot test Linux arm64 release build please

bgribaudo commented:
@dotet-bot test Linux arm64 release build please

bgribaudo commented:
Thanks, @danmosemsft, @JpratherMS, @riarenas, @ilyas1974 and @maririos for the help with the ARM64 timeout issue.

bgribaudo commented:
Two ideas that came while working on this for possible future consideration. Documenting them here for whatever they are worth (2 cents??):
* The way `Process` uses `Win32Exception` to report interop errors, it’s great for knowing that something bad happened but not always helpful in revealing _what the bad thing was_ that happened. Sometimes a native error code is included in the exception—but when it is, the interop method that returned the error code isn’t identified. It might be neat one day to extend this exception by adding a property that optionally holds the native (or platform abstraction layer) method name. This way, error codes returned can be associated with the context of what returned the error code. (Which is more helpful: an exception telling you that error code 6 was returned or one telling you that error code 6 was returned by `NtDll.NtQueryInformationProcess`?)
* On the Windows side, `Process` supports tracking a handle to the process structure it represents. When requested for an existing process (e.g. the `Process` instance was _not_ created by calling `Start`), the handle is [opened](https://github.com/dotnet/corefx/blob/master/src/System.Diagnostics.Process/src/System/Diagnostics/Process.cs#L1131) ([ref2](https://github.com/dotnet/corefx/blob/master/src/System.Diagnostics.Process/src/System/Diagnostics/Process.Windows.cs#L364-L367)) with `ALL_ACCESS` rights. It could be interesting to refactor this to instead use something like `PROCESS_QUERY_LIMITED_INFORMATION`. In the case of this PR, there’s a need to hold process handles but `ALL_ACCESS` is more permissive than appropriate so the PR implements its own handle management. If this refactoring were done, `Process`’s built-in handle tracking could be used instead.

bgribaudo commented:
Looks like the Linux arm64 Release Build issue has to do with a native (interop?) library issue not related to this PR. Maybe it will be cleared up in master soon.

MattGal commented:
@bgribaudo yes there was some sort of new libcurl3 dependency added which we've since added to the Ubuntu 18 ARM64 machines, should be fixed now.

bgribaudo commented:
Thanks, @MattGal!

danmosemsft commented on src/System.Diagnostics.Process/src/System/Diagnostics/Process.NonUap.cs:
```diff
@@ -0,0 +1,84 @@
+                    IEnumerable<Process> children = processes
+                        .Where(p => SafePredicateTest(() => current.IsParentOf(p)));
+
+                    if (children.Any(c => SafePredicateTest(() => c.Equals(processOfInterest))))
```
Would this be any clearer, and potentially do less work as it enumerates only once and potentially returns before filtering all the children?
```C#
do
{
	foreach(Process process in processes)
	{
     	if (SafePredicateTest(() -> current.IsParentOf(process))
     	{
         	if (SafePredicateTest(() => c.Equals(processOfInterest)))
              	return true;

	         queue.Enqueue(process)
    	}
	}
} while (queue.TryDequeue(out current));
```

danmosemsft commented on src/System.Diagnostics.Process/src/System/Diagnostics/Process.NonUap.cs:
```diff
@@ -0,0 +1,84 @@
+
+            try
+            {
+                var queue = new Queue<Process>();
```
is queue -> potentialChildren  any clearer?

danmosemsft commented:
@krwq @wtgodbe you are area owners of Process, can one of you please signoff.

@tmds does the treatment of Unix processes look correct?

Thanks for your patience @bgribaudo .. usually PR's go through a lot more quickly! I have a feeling we will discover more edge cases after this is committed.

krwq commented on src/System.Diagnostics.Process/src/System/Diagnostics/Process.NonUap.cs:
```diff
@@ -0,0 +1,84 @@
+            }
+        }
+
+        private bool IsInTreeOf(Process processOfInterest)
```
I'd call it IsDescendant/IsDescendantOf

krwq commented on src/System.Diagnostics.Process/src/System/Diagnostics/Process.NonUap.cs:
```diff
@@ -0,0 +1,84 @@
+
+                do
+                {
+                    IEnumerable<Process> children = processes
```
I might be missing something but name of this function suggests that there are 2 nodes:
- root (processOfInterest)
- potential descendant (this)
and that is suggested because the call looks like: `this.IsInTreeOf(root)` which I read as "is `this` a root or its descendant?"

The current implementation is following:
- we get a candidate which is either `this` or first element in the queue
- then a list of potential candidates is added with the children of this

```
      process of interest
          /\
         /  \
        /    \
       /\
      /  \
     /    \
          / this/current
         /
        /
       /\ next iteration (because current.IsParentOf(p) will be true)
      /  \
     /    \
```

I feel like we should be going up and next candidates should be parents (unless the function is misnamed)

am I missing anything here?

krwq replied:
But then on the other hand we check for `IsInTreeOf(GetCurrentProcess())` which again suggests what I wrote :-)

bgribaudo replied:
> Assuming the current logic is correct, can we somehow use GetChildProcesses here?

Technically, yes. That method could be called recursively for each child process to get its children, and the on each of those children to get their children, etc. The catch is that each time it's called, it fetches a fresh copy of the currently running processes list. 

A fresh copy of that list on each call is needed in order for GetChildProcesses to satisfy the need of its current consumer. However, in this section of code, it's safe to pull that list once and then scan the cached list multiple times to build out the descendant process tree.

If we're okay with paying the (small) cost of recursively calling GetChildProcesses, I'd be glad to change to doing that.

krwq replied:
Could we perhaps make GetChildProcesses take a list it uses and if it is i.e. null then we fetch it?

krwq replied:
I won't have time to carefully look at this PR today, probably will have a bit over the weekend or Monday (ping me if I didn't)

bgribaudo replied:
Reworked so that GetChildProcesses is used here.

krwq commented on src/System.Diagnostics.Process/src/System/Diagnostics/Process.NonUap.cs:
```diff
@@ -0,0 +1,84 @@
+                return false;
+            }
+        }
+
```
nit: extra enter

bgribaudo commented on src/System.Diagnostics.Process/src/System/Diagnostics/Process.NonUap.cs:
```diff
@@ -0,0 +1,84 @@
+
+                do
+                {
+                    IEnumerable<Process> children = processes
```
Thanks!

krwq commented on src/System.Diagnostics.Process/src/System/Diagnostics/Process.NonUap.cs:
```diff
@@ -0,0 +1,120 @@
+        /// </summary>
+        private IReadOnlyList<Process> GetChildProcesses(Process[] processes = null)
+        {
+            bool internallyInitializedProcesses = false;
```
nit: this could be:

```csharp
bool internallyInitializedProcesses = processes == null;
processes = processes ?? GetProcesses();
```

bgribaudo replied:
Thanks for the refactoring! Applied.

krwq commented on src/System.Diagnostics.Process/tests/Helpers.cs:
```diff
@@ -0,0 +1,28 @@
+﻿using System;
```
license header missing

bgribaudo replied:
Added. Thanks!

krwq commented on src/System.Diagnostics.Process/tests/ProcessTests.cs:
```diff
@@ -1007,7 +1007,7 @@ public void TestGetProcessById()
         {
             Process currentProcess = Process.GetCurrentProcess();
-
+            
```
nit: extra space

bgribaudo replied:
Fixed. Thanks!

krwq commented on src/System.Diagnostics.Process/tests/ProcessTests.netcoreapp.cs:
```diff
@@ -123,5 +129,264 @@ public void BothArgumentSetAndArgumentListSet()
+            var process = new Process();
+            Assert.Throws<InvalidOperationException>(() => process.Kill(entireProcessTree: true));
+        }
+
```
nit: extra enter

bgribaudo replied:
Fixed. Thanks!

krwq commented on src/System.Diagnostics.Process/tests/ProcessTests.netcoreapp.cs:
```diff
@@ -123,5 +129,264 @@ public void BothArgumentSetAndArgumentListSet()
+        public void Kill_EntireProcessTree_True_ProcessNotStarted_ThrowsInvalidOperationException()
+        {
+            var process = new Process();
+            Assert.Throws<InvalidOperationException>(() => process.Kill(entireProcessTree: true));
```
out of curiosity - not expecting any fixes - what's the issue on Uap?

bgribaudo replied:
I didn't implement on UAP because I thought UAP doesn't give access to parent process IDs. (I don't see an API to access the parent PID at https://docs.microsoft.com/en-us/uwp/win32-and-com/win32-apis.)

Does that sound correct or am I missing something? :-)

krwq replied:
Sounds good, I wouldn't worry about UAP yet, if we need to fix it we can fix it separately 😄

bgribaudo replied:
:-)

krwq commented on src/System.Diagnostics.Process/tests/ProcessTests.netcoreapp.cs:
```diff
@@ -123,5 +129,264 @@ public void BothArgumentSetAndArgumentListSet()
+            }
+
+            Assert.Throws<NotSupportedException>(() => process.Kill(entireProcessTree: true));
+
```
nit: extra enter

bgribaudo replied:
Thanks! Fixed.

krwq commented on src/System.Diagnostics.Process/tests/ProcessTests.netcoreapp.cs:
```diff
@@ -123,5 +129,264 @@ public void BothArgumentSetAndArgumentListSet()
+                parentProcess.WaitForExit();
+
+                return parentProcess.ExitCode;
+
```
ditto

krwq replied:
same comment for other places

bgribaudo replied:
Thanks! Fixed in several places! (Sorry about that!)

bgribaudo commented:
@dotnet-bot please test Windows x64 debug build

tmds commented on src/System.Diagnostics.Process/src/System/Diagnostics/Process.NonUap.cs:
```diff
@@ -0,0 +1,115 @@
+            }
+        }
+
+        private bool IsSelfOrDescendant(Process processOfInterest)
```
IsSelfOrParentOf ?

krwq replied:
Parent suggest imediate parent, descendant suggests the tree - this is the name XPath is using as well

krwq replied:
I don't mind adding `Of` though

bgribaudo replied:
Add `Of`--`IsSelfOrDescendantOf`

tmds commented on src/System.Diagnostics.Process/src/System/Diagnostics/Process.NonUap.cs:
```diff
@@ -0,0 +1,115 @@
+            {
+                return predicate();
+            }
+            catch (Exception e) when (e is InvalidOperationException || e is Win32Exception)
```
This is a very broad try catch.
An alternative would be to:
- not catch Win32Exceptions, e.g. permission issues should bubble up
- make ParentProcessId a nullable int, that is null when the process has exited.

danmosemsft replied:
It is also a bit friendlier for debugging if we try to not throw 1st chance exceptions in a "fairly normal" case like "already exited"

bgribaudo replied:
This bit of code is used only when determining which processes are children, not when performing process termination (so it's not suppressing exceptions encountered when attempting termination).

On the Windows side, a process doesn't necessarily have permission to access details on all other processes so encountering permissions errors can be a part of the normal course of business (e.g. a non-elevated process's attempting to get details on an elevated process may fail with a Win32Exception). In this context, if our goal is to make a best-effort attempt to terminate a process tree, it seems like these exceptions should be suppressed.

tmds replied:
makes sense

tmds commented on src/System.Diagnostics.Process/src/System/Diagnostics/Process.Unix.cs:
```diff
@@ -63,6 +64,69 @@ public void Kill()
+
+        private IEnumerable<Exception> KillTree()
+        {
+            List<Exception> exceptions = new List<Exception>();
```
It would be nice if the implementation would treat these cases the same:
Process instance knows the Process has exited
OS returns ESRCH because the process no longer exist
Currently one gives a InvalidOperationException (on HaveNonExitedId) that is swallowed and the other adds a Win32Exception exception to the list.

krwq replied:
I'd suggest to file issue for that (or perhaps 1 joint issue for all remaining feedback) and address it separately - this PR is already large enough as is.

bgribaudo replied:
Related: https://github.com/dotnet/corefx/issues/35257 -- if implemented, would resolve this issue.

tmds commented on src/System.Diagnostics.Process/src/System/Diagnostics/Process.Unix.cs:
```diff
@@ -63,6 +64,69 @@ public void Kill()
+
+        private IEnumerable<Exception> KillTree()
+        {
+            List<Exception> exceptions = new List<Exception>();
```
Instead of creating an aggregated list of exceptions, an alternative would be to stop on the first Process that causes an issue.
It may be simpler for the user to deal with a single exception.

bgribaudo replied:
I wonder if an optional argument could eventually be added to request this behavior (something like `throwOnFirstFailure: true`). 

The catch with doing it by default is that it leaves the consumer without an easy way to continue termination if what they really want is best-effort, try to terminate all descendants, regardless of whether exceptions are encountered. 

If termination dies on an exception partway through and the consumer wants execution to resume, they won't necessarily be able to re-call `Kill(entireProcessTree: true)` on the `Process` instance they started with because that instance may be dead. To resume, they'd have to figure out which process(es) is/are the orphan(s) left when termination died because of the exception, then call  `Kill(entireProcessTree: true)` on each of them. I don't think the consumer will have an easy way to find those orphans. 

This makes me hesitate with making _die on first exception_ the default behavior.

tmds commented on src/System.Diagnostics.Process/src/System/Diagnostics/Process.Unix.cs:
```diff
@@ -63,6 +64,69 @@ public void Kill()
+            {
+                // Stop but don't kill the process. Keeps additional children from being started but leaves the process alive 
+                // so that its children can be enumerated.
+                Stop();
```
When Stop returns, the process may be _stopping_ instead of being _stopped_.

bgribaudo replied:
Thanks! I didn't realize that. Is it only when stopping has completed that we are guaranteed the process can't start child processes? Is there an (easy) way to wait for stopping to complete?

tmds replied:
From the man page it seems you can do this with `waitid`. I'm not familiar with using that API and its quirks. Maybe you can add a comment about this.

bgribaudo replied:
Thanks! Comment added.

bgribaudo commented on src/System.Diagnostics.Process/tests/ProcessTests.netcoreapp.cs:
```diff
@@ -123,5 +129,264 @@ public void BothArgumentSetAndArgumentListSet()
+        [Fact]
+        [PlatformSpecific(TestPlatforms.Windows)]
+        [SkipOnTargetFramework(~TargetFrameworkMonikers.Uap)]
+        public void Kill_EntireProcessTree_ThrowsPlatformNotSupportedException()
```
@krwq, is [Kill_EntireProcessTree_True_CalledOnCallingProcess_ThrowsInvalidOperationException()](https://github.com/dotnet/corefx/pull/34147/files/e4e40b527a2b474b47bb6ee2565acc3bf7c8d678#diff-ff4583cbf53e8e40154f1232a26e17b7R178) what you're looking for?

krwq replied:
Must've missed it, yes, that is what I was looking for

bgribaudo commented on src/System.Diagnostics.Process/tests/Helpers.cs:
```diff
@@ -0,0 +1,28 @@
+                catch (XunitException) when (times > 1)
+                {
+                    await Task.Delay(delayInMilliseconds);
+                    delayInMilliseconds *= 2;
```
Added. I'm not sure what value to use for the cut-off, though. You mentioned maybe 1 second; @danmosemsft [previously suggested](https://github.com/dotnet/corefx/pull/31827#discussion_r213510986) allowing things to go a bit longer. Right now, I have it set to 10 seconds. Glad to adjust to more or less if that is not a good compromise value.

danmosemsft replied:
I suggest to leave as is, I have increased so many timeouts on tests that for some reason run slow in official runs 1 in 1000 times because the machine is paging or something.

bgribaudo replied:
Sounds good to me.

bgribaudo commented on src/System.Diagnostics.Process/src/System/Diagnostics/Process.FreeBSD.cs:
```diff
@@ -63,11 +63,15 @@ public TimeSpan PrivilegedProcessorTime
 
+        /// <summary>Gets parent process ID</summary>
+        private int ParentProcessId =>
+            throw new PlatformNotSupportedException();
```
By issue, do you mean a GitHub issue?

danmosemsft replied:
Yep. Some community member may be interested in filling this hole

bgribaudo replied:
I'll plan to do that once this one goes through.

bgribaudo replied:
Issue created: https://github.com/dotnet/corefx/issues/35290

bgribaudo commented on src/System.Diagnostics.Process/src/System/Diagnostics/Process.Unix.cs:
```diff
@@ -63,6 +64,69 @@ public void Kill()
+            }
+            catch (Win32Exception e)
+            {
+                exceptions.Add(e);
```
I guess not. :-) Thanks, @tmds! Fixed.

bgribaudo commented:
@dotnet-bot please test Windows x64 Debug Build

tmds commented on src/System.Diagnostics.Process/src/System/Diagnostics/Process.Win32.cs:
```diff
@@ -345,5 +347,108 @@ private bool WaitForInputIdleCore(int milliseconds)
+                // Kill the process, so that no further children can be created.
+                //
+                // This method can return before stopping has completed. Down the road, could possibly wait for termination to complete before continuing.
+                Kill();
```
Maybe we should do this already, calling WaitForExit should be enough.

danmosemsft replied:
I do not know. To be certain it is probably necessary to wait on the process handle to be signaled (which means it's finished terminating) and pessimistically assume it could have spawned more processes. That seems a corner case we can perhaps ignore for now.

bgribaudo replied:
Thanks, @tmds & @dnfadmin. Sounds like one of those many edge cases Dan mentioned that we'd discover.... :-)

danmosemsft commented:
@dotnet-bot test NETFX x86 Release Build please (AV in csc .. bit worrying, but no diagnostics)

danmosemsft commented:
@wtgodbe do you have any further feedback?

safern commented:
This branch still has an old Helix SDK, that is why we're hitting the RestApiException. I'll retry the failing leg.

wtgodbe commented on src/System.Diagnostics.Process/src/System/Diagnostics/Process.NonUap.cs:
```diff
@@ -0,0 +1,115 @@
+
+                        descendantProcesses.Enqueue(candidate);
+                    }
+                } while (descendantProcesses.TryDequeue(out current));
```
nit: repo convention see that the `while` should be on its own line below the `}`. https://github.com/dotnet/corefx/blob/master/src/Common/src/CoreLib/System/Buffers/ConfigurableArrayPool.cs#L86-L99

danmosemsft replied:
I did a regex and there's 481 on the same line and 600 on separate lines. I call that a wash 😃

wtgodbe commented on src/System.Diagnostics.Process/tests/Helpers.cs:
```diff
@@ -0,0 +1,37 @@
+    {
+        public static async Task RetryWithBackoff(Action action, int delayInMilliseconds = 10, int times = 10)
+        {
+            const int maxDelayInMilliseconds = 10000;
```
Why 10000? Maybe add a comment here?

bgribaudo replied:
Good idea. Comment added.

danmosemsft commented:
@tmds any remaining feedback?

bgribaudo commented:
@dotnet-bot please test corefx-ci

bgribaudo commented on src/System.Diagnostics.Process/src/System/Diagnostics/Process.Unix.cs:
```diff
@@ -240,6 +304,42 @@ private static int GetCurrentProcessId()
+        }
+
+        /// <summary>Checks whether the argument is a direct child of this process.</summary>
+        private bool IsParentOf(Process possibleChildProcess) =>
```
@danmosemsft, is this something you'd really like done or just an idea? :-) 

I am glad to make this change if you want it done. If it's just an idea where you're okay with going either way on it, I wouldn't mind sticking with `IsParentOf` as that terminology matches what's used elsewhere in this PR (e.g. we ask the OS for the parent process ID/ppid, not the immediate parent process ID/ippid).

All this said, if you want the change made, just say so and it will be done.

danmosemsft replied:
Fine to leave as is :)

bgribaudo replied:
Thanks! Hope you have a great weekend!

safern commented:
Note that to trigger corefx-ci comments are broken. It is a known issue and we're waiting for a fix.

It is probable that the build will fail. It seems like if you want it working, you will need to rebase or merge with master. Since all other builds are green and corefx-ci covers the same builds. We can ignore it in this PR since it is fairly old.

danmosemsft commented:
This has been open long enough and scrutinized enough 😃 

Thanks a lot @bgribaudo for the contribution. As you can see, usually community PR's go through much faster. I guess it was several reasons including the fiddly change. I hope it does not deter you from contributing in future. We have lots of issues labeled up for grabs.

krwq commented:
Thanks @bgribaudo

bgribaudo commented:
Thank you, @danmosemsft, @krwq, @tmds, @wtgodbe (and anyone else I'm missing that participated)! I appreciate your patience on this one. :-)
